virtio-queue/src/descriptor_utils.rs

```diff
@@ -0,0 +1,884 @@
+
+        // This can happen if a driver tricks a device into reading/writing more data than
+        // fits in a `usize`.
+        let total_bytes_consumed =
+            self.bytes_consumed
+                .checked_add(bytes_consumed)
+                .ok_or_else(|| {
+                    io::Error::new(io::ErrorKind::InvalidData, Error::DescriptorChainOverflow)
+                })?;
+
+        let mut rem = bytes_consumed;
+        while let Some(vs) = self.buffers.pop_front() {
+            if rem < vs.len() {
+                // Split the slice and push the remainder back into the buffer list. Safe because we
+                // know that `rem` is not out of bounds due to the check and we checked the bounds
+                // on `vs` when we added it to the buffer list.
+                self.buffers.push_front(vs.offset(rem).unwrap());
+                break;
+            }
+
+            // No need for checked math because we know that `vs.size() <= rem`.
+            rem -= vs.len();
+        }
+
+        self.bytes_consumed = total_bytes_consumed;
+
+        Ok(bytes_consumed)
+    }
+
+    fn split_at(&mut self, offset: usize) -> Result<DescriptorChainConsumer<'a, B>> {
+        let mut rem = offset;
+        let pos = self.buffers.iter().position(|vs| {
+            if rem < vs.len() {
+                true
+            } else {
+                rem -= vs.len();
+                false
+            }
+        });
+
+        if let Some(at) = pos {
+            let mut other = self.buffers.split_off(at);
+
+            if rem > 0 {
+                // There must be at least one element in `other` because we checked
+                // its `size` value in the call to `position` above.
+                let front = other.pop_front().expect("empty VecDeque after split");
+                self.buffers
+                    .push_back(front.subslice(0, rem).map_err(Error::VolatileMemoryError)?);
+                other.push_front(front.offset(rem).map_err(Error::VolatileMemoryError)?);
+            }
+
+            Ok(DescriptorChainConsumer {
+                buffers: other,
+                bytes_consumed: 0,
+            })
+        } else if rem == 0 {
+            Ok(DescriptorChainConsumer {
+                buffers: VecDeque::new(),
+                bytes_consumed: 0,
+            })
+        } else {
+            Err(Error::SplitOutOfBounds(offset))
+        }
+    }
+}
+
+/// Provides high-level interface over the sequence of memory regions
+/// defined by readable descriptors in the descriptor chain.
+///
+/// Note that virtio spec requires driver to place any device-writable
+/// descriptors after any device-readable descriptors (2.6.4.2 in Virtio Spec v1.1).
+/// Reader will skip iterating over descriptor chain when first writable
+/// descriptor is encountered.
+#[derive(Clone)]
+pub struct Reader<'a, B = ()> {
+    buffer: DescriptorChainConsumer<'a, B>,
+}
+
+impl<'a, B: Bitmap + BitmapSlice + 'static> Reader<'a, B> {
```

Comment:
why we need `'static` here?

Reply:
Short answer: Because abstractions in vm-memory are broken.

Long answer: The slack discussion on this is currently still readable: https://rust-vmm.slack.com/archives/CFH3R8PFV/p1701337335710629.
Summary: vm-memory uses higher-ranked-traits to seemingly cheat out of needing to leak lifetimes into the GuestMemory trait. But in reality this basically just causes everything needing 'static lifetime. I tried to fix it when @MatiasVara posted this PR, but the refactoring attempt exploded and started leaking lifetimes into a lot of public interfaces: https://github.com/Ablu/vm-memory/commit/b22bdbdc953778b1693de6c95044beda4293d05c.

Reply:
@Ablu I see, but for example we can remove `Bitmap` since it is implicit in `BitmapSlice`, no?

BTW in the patch I posted I left only BitmapSlice and the compiler seems happy, I need to investigate that more.

Reply:
> @Ablu I see, but for example we can remove `Bitmap` since it is implicit in `BitmapSlice`, no?
> 
> BTW in the patch I posted I left only BitmapSlice and the compiler seems happy, I need to investigate that more.

Okay, I can do that in my patch because of setting the lifetime via WithBitmapSlice. I'm resolving this conversation.

Reply:
Yes, BitmapSlice is just a "special" Bitmap version. So we can drop the `Bitmap`.